// angular2-lightbox/src/lightbox.component.ngfactory.ts
/**
 * @fileoverview This file is generated by the Angular 2 template compiler.
 * Do not edit.
 * @suppress {suspiciousCode,uselessCode,missingProperties}
 */
 /* tslint:disable */

import * as import0 from './lightbox.component';
import * as import1 from '@angular/core/src/change_detection/change_detection_util';
import * as import2 from '@angular/core/src/linker/view';
import * as import3 from '@angular/core/src/linker/view_utils';
import * as import4 from '@angular/core/src/render/api';
import * as import5 from '@angular/core/src/metadata/view';
import * as import6 from '@angular/core/src/linker/view_type';
import * as import7 from '@angular/core/src/change_detection/constants';
import * as import8 from '@angular/core/src/linker/component_factory';
import * as import9 from '@angular/core/src/linker/element_ref';
import * as import10 from './lightbox-event.service';
import * as import11 from '@angular/platform-browser/src/security/dom_sanitization_service';
import * as import12 from '@angular/platform-browser/src/dom/dom_tokens';
import * as import13 from '@angular/core/src/linker/query_list';
import * as import14 from '@angular/core/src/security';
export class Wrapper_LightboxComponent {
  /*private*/ _eventHandler:Function;
  context:import0.LightboxComponent;
  /*private*/ _changed:boolean;
  /*private*/ _expr_0:any;
  /*private*/ _expr_1:any;
  /*private*/ _expr_2:any;
  /*private*/ _expr_3:any;
  /*private*/ _expr_4:any;
  constructor(p0:any,p1:any,p2:any,p3:any,p4:any,p5:any,p6:any) {
    this._changed = false;
    this.context = new import0.LightboxComponent(p0,p1,p2,p3,p4,p5,p6);
    this._expr_0 = import1.UNINITIALIZED;
    this._expr_1 = import1.UNINITIALIZED;
    this._expr_2 = import1.UNINITIALIZED;
    this._expr_3 = import1.UNINITIALIZED;
    this._expr_4 = import1.UNINITIALIZED;
  }
  ngOnDetach(view:import2.AppView<any>,componentView:import2.AppView<any>,el:any):void {
  }
  ngOnDestroy():void {
    this.context.ngOnDestroy();
  }
  check_album(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import3.checkBinding(throwOnChange,this._expr_0,currValue))) {
      this._changed = true;
      this.context.album = currValue;
      this._expr_0 = currValue;
    }
  }
  check_currentImageIndex(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import3.checkBinding(throwOnChange,this._expr_1,currValue))) {
      this._changed = true;
      this.context.currentImageIndex = currValue;
      this._expr_1 = currValue;
    }
  }
  check_options(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import3.checkBinding(throwOnChange,this._expr_2,currValue))) {
      this._changed = true;
      this.context.options = currValue;
      this._expr_2 = currValue;
    }
  }
  check_cmpRef(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import3.checkBinding(throwOnChange,this._expr_3,currValue))) {
      this._changed = true;
      this.context.cmpRef = currValue;
      this._expr_3 = currValue;
    }
  }
  ngDoCheck(view:import2.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this._changed;
    this._changed = false;
    if (!throwOnChange) { if ((view.numberOfChecks === 0)) { this.context.ngOnInit(); } }
    return changed;
  }
  checkHost(view:import2.AppView<any>,componentView:import2.AppView<any>,el:any,throwOnChange:boolean):void {
    const currVal_4:any = this.context.ui.classList;
    if (import3.checkBinding(throwOnChange,this._expr_4,currVal_4)) {
      view.renderer.setElementProperty(el,'className',currVal_4);
      this._expr_4 = currVal_4;
    }
  }
  handleEvent(eventName:string,$event:any):boolean {
    var result:boolean = true;
    if ((eventName == 'click')) {
      const pd_sub_0:any = ((<any>this.context.close($event)) !== false);
      result = (pd_sub_0 && result);
    }
    return result;
  }
  subscribe(view:import2.AppView<any>,_eventHandler:any):void {
    this._eventHandler = _eventHandler;
  }
}
var renderType_LightboxComponent_Host:import4.RenderComponentType = import3.createRenderComponentType('',0,import5.ViewEncapsulation.None,([] as any[]),{});
class View_LightboxComponent_Host0 extends import2.AppView<any> {
  _el_0:any;
  compView_0:import2.AppView<import0.LightboxComponent>;
  _LightboxComponent_0_3:Wrapper_LightboxComponent;
  constructor(viewUtils:import3.ViewUtils,parentView:import2.AppView<any>,parentIndex:number,parentElement:any) {
    super(View_LightboxComponent_Host0,renderType_LightboxComponent_Host,import6.ViewType.HOST,viewUtils,parentView,parentIndex,parentElement,import7.ChangeDetectorStatus.CheckAlways);
  }
  createInternal(rootSelector:string):import8.ComponentRef<any> {
    this._el_0 = import3.selectOrCreateRenderHostElement(this.renderer,'div',new import3.InlineArray2(2,'lb-content',''),rootSelector,(null as any));
    this.compView_0 = new View_LightboxComponent0(this.viewUtils,this,0,this._el_0);
    this._LightboxComponent_0_3 = new Wrapper_LightboxComponent(new import9.ElementRef(this._el_0),this.renderer,this.injectorGet(import10.LightboxEvent,this.parentIndex),new import9.ElementRef(this._el_0),this.injectorGet(import10.LightboxWindowRef,this.parentIndex),this.injectorGet(import11.DomSanitizer,this.parentIndex),this.injectorGet(import12.DOCUMENT,this.parentIndex));
    this.compView_0.create(this._LightboxComponent_0_3.context);
    var disposable_0:Function = import3.subscribeToRenderElement(this,this._el_0,new import3.InlineArray2(2,'click',(null as any)),this.eventHandler(this.handleEvent_0));
    this.init(this._el_0,((<any>this.renderer).directRenderer? (null as any): [this._el_0]),[disposable_0]);
    return new import8.ComponentRef_<any>(0,this,this._el_0,this._LightboxComponent_0_3.context);
  }
  injectorGetInternal(token:any,requestNodeIndex:number,notFoundResult:any):any {
    if (((token === import0.LightboxComponent) && (0 === requestNodeIndex))) { return this._LightboxComponent_0_3.context; }
    return notFoundResult;
  }
  detectChangesInternal(throwOnChange:boolean):void {
    this._LightboxComponent_0_3.ngDoCheck(this,this._el_0,throwOnChange);
    this._LightboxComponent_0_3.checkHost(this,this.compView_0,this._el_0,throwOnChange);
    this.compView_0.internalDetectChanges(throwOnChange);
    if (!throwOnChange) { if ((this.numberOfChecks === 0)) { this._LightboxComponent_0_3.context.ngAfterViewInit(); } }
  }
  destroyInternal():void {
    this.compView_0.destroy();
    this._LightboxComponent_0_3.ngOnDestroy();
  }
  visitRootNodesInternal(cb:any,ctx:any):void {
    cb(this._el_0,ctx);
  }
  handleEvent_0(eventName:string,$event:any):boolean {
    this.compView_0.markPathToRootAsCheckOnce();
    var result:boolean = true;
    result = (this._LightboxComponent_0_3.handleEvent(eventName,$event) && result);
    return result;
  }
}
export const LightboxComponentNgFactory:import8.ComponentFactory<import0.LightboxComponent> = new import8.ComponentFactory<import0.LightboxComponent>('[lb-content]',View_LightboxComponent_Host0,import0.LightboxComponent);
const styles_LightboxComponent:any[] = ([] as any[]);
var renderType_LightboxComponent:import4.RenderComponentType = import3.createRenderComponentType('',0,import5.ViewEncapsulation.None,styles_LightboxComponent,{});
export class View_LightboxComponent0 extends import2.AppView<import0.LightboxComponent> {
  _viewQuery_outerContainer_0:import13.QueryList<any>;
  _viewQuery_container_1:import13.QueryList<any>;
  _viewQuery_leftArrow_2:import13.QueryList<any>;
  _viewQuery_rightArrow_3:import13.QueryList<any>;
  _viewQuery_navArrow_4:import13.QueryList<any>;
  _viewQuery_dataContainer_5:import13.QueryList<any>;
  _viewQuery_image_6:import13.QueryList<any>;
  _viewQuery_caption_7:import13.QueryList<any>;
  _viewQuery_number_8:import13.QueryList<any>;
  _text_0:any;
  _el_1:any;
  _text_2:any;
  _el_3:any;
  _text_4:any;
  _el_5:any;
  _text_6:any;
  _el_7:any;
  _text_8:any;
  _el_9:any;
  _text_10:any;
  _el_11:any;
  _text_12:any;
  _text_13:any;
  _el_14:any;
  _text_15:any;
  _el_16:any;
  _text_17:any;
  _text_18:any;
  _text_19:any;
  _text_20:any;
  _el_21:any;
  _text_22:any;
  _el_23:any;
  _text_24:any;
  _el_25:any;
  _text_26:any;
  _el_27:any;
  _text_28:any;
  _text_29:any;
  _el_30:any;
  _text_31:any;
  _text_32:any;
  _text_33:any;
  _el_34:any;
  _text_35:any;
  _el_36:any;
  _text_37:any;
  _text_38:any;
  _text_39:any;
  /*private*/ _expr_49:any;
  /*private*/ _expr_50:any;
  /*private*/ _expr_51:any;
  /*private*/ _expr_52:any;
  /*private*/ _expr_53:any;
  /*private*/ _expr_54:any;
  /*private*/ _expr_55:any;
  /*private*/ _expr_56:any;
  /*private*/ _expr_57:any;
  /*private*/ _expr_58:any;
  /*private*/ _expr_59:any;
  constructor(viewUtils:import3.ViewUtils,parentView:import2.AppView<any>,parentIndex:number,parentElement:any) {
    super(View_LightboxComponent0,renderType_LightboxComponent,import6.ViewType.COMPONENT,viewUtils,parentView,parentIndex,parentElement,import7.ChangeDetectorStatus.CheckAlways);
    this._expr_49 = import1.UNINITIALIZED;
    this._expr_50 = import1.UNINITIALIZED;
    this._expr_51 = import1.UNINITIALIZED;
    this._expr_52 = import1.UNINITIALIZED;
    this._expr_53 = import1.UNINITIALIZED;
    this._expr_54 = import1.UNINITIALIZED;
    this._expr_55 = import1.UNINITIALIZED;
    this._expr_56 = import1.UNINITIALIZED;
    this._expr_57 = import1.UNINITIALIZED;
    this._expr_58 = import1.UNINITIALIZED;
    this._expr_59 = import1.UNINITIALIZED;
  }
  createInternal(rootSelector:string):import8.ComponentRef<any> {
    const parentRenderNode:any = this.renderer.createViewRoot(this.parentElement);
    this._viewQuery_outerContainer_0 = new import13.QueryList<any>();
    this._viewQuery_container_1 = new import13.QueryList<any>();
    this._viewQuery_leftArrow_2 = new import13.QueryList<any>();
    this._viewQuery_rightArrow_3 = new import13.QueryList<any>();
    this._viewQuery_navArrow_4 = new import13.QueryList<any>();
    this._viewQuery_dataContainer_5 = new import13.QueryList<any>();
    this._viewQuery_image_6 = new import13.QueryList<any>();
    this._viewQuery_caption_7 = new import13.QueryList<any>();
    this._viewQuery_number_8 = new import13.QueryList<any>();
    this._text_0 = this.renderer.createText(parentRenderNode,'\n    ',(null as any));
    this._el_1 = import3.createRenderElement(this.renderer,parentRenderNode,'div',new import3.InlineArray2(2,'class','lb-outerContainer transition'),(null as any));
    this._text_2 = this.renderer.createText(this._el_1,'\n      ',(null as any));
    this._el_3 = import3.createRenderElement(this.renderer,this._el_1,'div',new import3.InlineArray2(2,'class','lb-container'),(null as any));
    this._text_4 = this.renderer.createText(this._el_3,'\n        ',(null as any));
    this._el_5 = import3.createRenderElement(this.renderer,this._el_3,'img',new import3.InlineArray2(2,'class','lb-image animation fadeIn'),(null as any));
    this._text_6 = this.renderer.createText(this._el_3,'\n        ',(null as any));
    this._el_7 = import3.createRenderElement(this.renderer,this._el_3,'div',new import3.InlineArray2(2,'class','lb-nav'),(null as any));
    this._text_8 = this.renderer.createText(this._el_7,'\n          ',(null as any));
    this._el_9 = import3.createRenderElement(this.renderer,this._el_7,'a',new import3.InlineArray2(2,'class','lb-prev'),(null as any));
    this._text_10 = this.renderer.createText(this._el_7,'\n          ',(null as any));
    this._el_11 = import3.createRenderElement(this.renderer,this._el_7,'a',new import3.InlineArray2(2,'class','lb-next'),(null as any));
    this._text_12 = this.renderer.createText(this._el_7,'\n        ',(null as any));
    this._text_13 = this.renderer.createText(this._el_3,'\n        ',(null as any));
    this._el_14 = import3.createRenderElement(this.renderer,this._el_3,'div',new import3.InlineArray2(2,'class','lb-loader'),(null as any));
    this._text_15 = this.renderer.createText(this._el_14,'\n          ',(null as any));
    this._el_16 = import3.createRenderElement(this.renderer,this._el_14,'a',new import3.InlineArray2(2,'class','lb-cancel'),(null as any));
    this._text_17 = this.renderer.createText(this._el_14,'\n        ',(null as any));
    this._text_18 = this.renderer.createText(this._el_3,'\n      ',(null as any));
    this._text_19 = this.renderer.createText(this._el_1,'\n    ',(null as any));
    this._text_20 = this.renderer.createText(parentRenderNode,'\n    ',(null as any));
    this._el_21 = import3.createRenderElement(this.renderer,parentRenderNode,'div',new import3.InlineArray2(2,'class','lb-dataContainer'),(null as any));
    this._text_22 = this.renderer.createText(this._el_21,'\n      ',(null as any));
    this._el_23 = import3.createRenderElement(this.renderer,this._el_21,'div',new import3.InlineArray2(2,'class','lb-data'),(null as any));
    this._text_24 = this.renderer.createText(this._el_23,'\n        ',(null as any));
    this._el_25 = import3.createRenderElement(this.renderer,this._el_23,'div',new import3.InlineArray2(2,'class','lb-details'),(null as any));
    this._text_26 = this.renderer.createText(this._el_25,'\n          ',(null as any));
    this._el_27 = import3.createRenderElement(this.renderer,this._el_25,'span',new import3.InlineArray2(2,'class','lb-caption animation fadeIn'),(null as any));
    this._text_28 = this.renderer.createText(this._el_27,'\n          ',(null as any));
    this._text_29 = this.renderer.createText(this._el_25,'\n          ',(null as any));
    this._el_30 = import3.createRenderElement(this.renderer,this._el_25,'span',new import3.InlineArray2(2,'class','lb-number animation fadeIn'),(null as any));
    this._text_31 = this.renderer.createText(this._el_30,'',(null as any));
    this._text_32 = this.renderer.createText(this._el_25,'\n        ',(null as any));
    this._text_33 = this.renderer.createText(this._el_23,'\n        ',(null as any));
    this._el_34 = import3.createRenderElement(this.renderer,this._el_23,'div',new import3.InlineArray2(2,'class','lb-closeContainer'),(null as any));
    this._text_35 = this.renderer.createText(this._el_34,'\n          ',(null as any));
    this._el_36 = import3.createRenderElement(this.renderer,this._el_34,'a',new import3.InlineArray2(2,'class','lb-close'),(null as any));
    this._text_37 = this.renderer.createText(this._el_34,'\n        ',(null as any));
    this._text_38 = this.renderer.createText(this._el_23,'\n      ',(null as any));
    this._text_39 = this.renderer.createText(this._el_21,'\n    ',(null as any));
    var disposable_0:Function = import3.subscribeToRenderElement(this,this._el_9,new import3.InlineArray2(2,'click',(null as any)),this.eventHandler(this.handleEvent_9));
    var disposable_1:Function = import3.subscribeToRenderElement(this,this._el_11,new import3.InlineArray2(2,'click',(null as any)),this.eventHandler(this.handleEvent_11));
    var disposable_2:Function = import3.subscribeToRenderElement(this,this._el_14,new import3.InlineArray2(2,'click',(null as any)),this.eventHandler(this.handleEvent_14));
    var disposable_3:Function = import3.subscribeToRenderElement(this,this._el_36,new import3.InlineArray2(2,'click',(null as any)),this.eventHandler(this.handleEvent_36));
    this._viewQuery_outerContainer_0.reset([new import9.ElementRef(this._el_1)]);
    this.context._outerContainerElem = this._viewQuery_outerContainer_0.first;
    this._viewQuery_container_1.reset([new import9.ElementRef(this._el_3)]);
    this.context._containerElem = this._viewQuery_container_1.first;
    this._viewQuery_leftArrow_2.reset([new import9.ElementRef(this._el_9)]);
    this.context._leftArrowElem = this._viewQuery_leftArrow_2.first;
    this._viewQuery_rightArrow_3.reset([new import9.ElementRef(this._el_11)]);
    this.context._rightArrowElem = this._viewQuery_rightArrow_3.first;
    this._viewQuery_navArrow_4.reset([new import9.ElementRef(this._el_7)]);
    this.context._navArrowElem = this._viewQuery_navArrow_4.first;
    this._viewQuery_dataContainer_5.reset([new import9.ElementRef(this._el_21)]);
    this.context._dataContainerElem = this._viewQuery_dataContainer_5.first;
    this._viewQuery_image_6.reset([new import9.ElementRef(this._el_5)]);
    this.context._imageElem = this._viewQuery_image_6.first;
    this._viewQuery_caption_7.reset([new import9.ElementRef(this._el_27)]);
    this.context._captionElem = this._viewQuery_caption_7.first;
    this._viewQuery_number_8.reset([new import9.ElementRef(this._el_30)]);
    this.context._numberElem = this._viewQuery_number_8.first;
    this.init((null as any),((<any>this.renderer).directRenderer? (null as any): [
      this._text_0,
      this._el_1,
      this._text_2,
      this._el_3,
      this._text_4,
      this._el_5,
      this._text_6,
      this._el_7,
      this._text_8,
      this._el_9,
      this._text_10,
      this._el_11,
      this._text_12,
      this._text_13,
      this._el_14,
      this._text_15,
      this._el_16,
      this._text_17,
      this._text_18,
      this._text_19,
      this._text_20,
      this._el_21,
      this._text_22,
      this._el_23,
      this._text_24,
      this._el_25,
      this._text_26,
      this._el_27,
      this._text_28,
      this._text_29,
      this._el_30,
      this._text_31,
      this._text_32,
      this._text_33,
      this._el_34,
      this._text_35,
      this._el_36,
      this._text_37,
      this._text_38,
      this._text_39
    ]
    ),[
      disposable_0,
      disposable_1,
      disposable_2,
      disposable_3
    ]
    );
    return (null as any);
  }
  detectChangesInternal(throwOnChange:boolean):void {
    const currVal_49:any = this.context.album[this.context.currentImageIndex].src;
    if (import3.checkBinding(throwOnChange,this._expr_49,currVal_49)) {
      this.renderer.setElementProperty(this._el_5,'src',this.viewUtils.sanitizer.sanitize(import14.SecurityContext.URL,currVal_49));
      this._expr_49 = currVal_49;
    }
    const currVal_50:any = this.context.ui.showReloader;
    if (import3.checkBinding(throwOnChange,this._expr_50,currVal_50)) {
      this.renderer.setElementProperty(this._el_5,'hidden',currVal_50);
      this._expr_50 = currVal_50;
    }
    const currVal_51:boolean = !this.context.ui.showArrowNav;
    if (import3.checkBinding(throwOnChange,this._expr_51,currVal_51)) {
      this.renderer.setElementProperty(this._el_7,'hidden',currVal_51);
      this._expr_51 = currVal_51;
    }
    const currVal_52:boolean = !this.context.ui.showLeftArrow;
    if (import3.checkBinding(throwOnChange,this._expr_52,currVal_52)) {
      this.renderer.setElementProperty(this._el_9,'hidden',currVal_52);
      this._expr_52 = currVal_52;
    }
    const currVal_53:boolean = !this.context.ui.showRightArrow;
    if (import3.checkBinding(throwOnChange,this._expr_53,currVal_53)) {
      this.renderer.setElementProperty(this._el_11,'hidden',currVal_53);
      this._expr_53 = currVal_53;
    }
    const currVal_54:boolean = !this.context.ui.showReloader;
    if (import3.checkBinding(throwOnChange,this._expr_54,currVal_54)) {
      this.renderer.setElementProperty(this._el_14,'hidden',currVal_54);
      this._expr_54 = currVal_54;
    }
    const currVal_55:any = this.context.ui.showReloader;
    if (import3.checkBinding(throwOnChange,this._expr_55,currVal_55)) {
      this.renderer.setElementProperty(this._el_21,'hidden',currVal_55);
      this._expr_55 = currVal_55;
    }
    const currVal_56:boolean = !this.context.ui.showCaption;
    if (import3.checkBinding(throwOnChange,this._expr_56,currVal_56)) {
      this.renderer.setElementProperty(this._el_27,'hidden',currVal_56);
      this._expr_56 = currVal_56;
    }
    const currVal_57:any = this.context.album[this.context.currentImageIndex].caption;
    if (import3.checkBinding(throwOnChange,this._expr_57,currVal_57)) {
      this.renderer.setElementProperty(this._el_27,'innerHTML',this.viewUtils.sanitizer.sanitize(import14.SecurityContext.HTML,currVal_57));
      this._expr_57 = currVal_57;
    }
    const currVal_58:boolean = !this.context.ui.showPageNumber;
    if (import3.checkBinding(throwOnChange,this._expr_58,currVal_58)) {
      this.renderer.setElementProperty(this._el_30,'hidden',currVal_58);
      this._expr_58 = currVal_58;
    }
    const currVal_59:any = import3.inlineInterpolate(1,'',this.context.content.pageNumber,'');
    if (import3.checkBinding(throwOnChange,this._expr_59,currVal_59)) {
      this.renderer.setText(this._text_31,currVal_59);
      this._expr_59 = currVal_59;
    }
  }
  handleEvent_9(eventName:string,$event:any):boolean {
    this.markPathToRootAsCheckOnce();
    var result:boolean = true;
    if ((eventName == 'click')) {
      const pd_sub_0:any = ((<any>this.context.prevImage()) !== false);
      result = (pd_sub_0 && result);
    }
    return result;
  }
  handleEvent_11(eventName:string,$event:any):boolean {
    this.markPathToRootAsCheckOnce();
    var result:boolean = true;
    if ((eventName == 'click')) {
      const pd_sub_0:any = ((<any>this.context.nextImage()) !== false);
      result = (pd_sub_0 && result);
    }
    return result;
  }
  handleEvent_14(eventName:string,$event:any):boolean {
    this.markPathToRootAsCheckOnce();
    var result:boolean = true;
    if ((eventName == 'click')) {
      const pd_sub_0:any = ((<any>this.context.close($event)) !== false);
      result = (pd_sub_0 && result);
    }
    return result;
  }
  handleEvent_36(eventName:string,$event:any):boolean {
    this.markPathToRootAsCheckOnce();
    var result:boolean = true;
    if ((eventName == 'click')) {
      const pd_sub_0:any = ((<any>this.context.close($event)) !== false);
      result = (pd_sub_0 && result);
    }
    return result;
  }
}
