// angulartics2/__ivy_ngcc__/routerlessmodule/fesm5/angulartics2-routerlessmodule.js
import { __decorate } from 'tslib';
import { NgModule } from '@angular/core';
import { ANGULARTICS2_TOKEN, RouterlessTracking, Angulartics2, Angulartics2OnModule } from 'angulartics2';

import * as ɵngcc0 from '@angular/core';
var Angulartics2RouterlessModule = /** @class */ (function () {
    function Angulartics2RouterlessModule() {
    }
    Angulartics2RouterlessModule_1 = Angulartics2RouterlessModule;
    Angulartics2RouterlessModule.forRoot = function (settings) {
        if (settings === void 0) { settings = {}; }
        return {
            ngModule: Angulartics2RouterlessModule_1,
            providers: [
                { provide: ANGULARTICS2_TOKEN, useValue: { settings: settings } },
                RouterlessTracking,
                Angulartics2,
            ],
        };
    };
    var Angulartics2RouterlessModule_1;
Angulartics2RouterlessModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Angulartics2RouterlessModule });
Angulartics2RouterlessModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Angulartics2RouterlessModule_Factory(t) { return new (t || Angulartics2RouterlessModule)(); }, imports: [[Angulartics2OnModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Angulartics2RouterlessModule, { imports: function () { return [Angulartics2OnModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Angulartics2RouterlessModule, [{
        type: NgModule,
        args: [{
                imports: [Angulartics2OnModule]
            }]
    }], function () { return []; }, null); })();
    return Angulartics2RouterlessModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { Angulartics2RouterlessModule };
