// angular2-lightbox/compiled/src/lightbox.component.ngfactory.ts
/**
 * @fileoverview This file is generated by the Angular template compiler.
 * Do not edit.
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride}
 */
 /* tslint:disable */


import * as import0 from '@angular/core';
import * as import1 from '../../src/lightbox.component';
import * as import2 from '../../src/lightbox-event.service';
import * as import3 from '@angular/platform-browser';
const styles_LightboxComponent:any[] = ([] as any[]);
export const RenderType_LightboxComponent:import0.RendererType2 = import0.ɵcrt({
  encapsulation: 2,
  styles: styles_LightboxComponent,
  data: {}
}
);
export function View_LightboxComponent_0(l:any):import0.ɵViewDefinition {
  return import0.ɵvid(0,[
    import0.ɵqud(201326592,1,{_outerContainerElem: 0}),
    import0.ɵqud(201326592,2,{_containerElem: 0}),
    import0.ɵqud(201326592,3,{_leftArrowElem: 0}),
    import0.ɵqud(201326592,4,{_rightArrowElem: 0}),
    import0.ɵqud(201326592,5,{_navArrowElem: 0}),
    import0.ɵqud(201326592,6,{_dataContainerElem: 0}),
    import0.ɵqud(201326592,7,{_imageElem: 0}),
    import0.ɵqud(201326592,8,{_captionElem: 0}),
    import0.ɵqud(201326592,9,{_numberElem: 0}),
    (l()(),import0.ɵted((null as any),['\n    '])),
    (l()(),import0.ɵeld(0,[
      [
        1,
        0
      ]
      ,
      [
        'outerContainer',
        1
      ]

    ]
      ,(null as any),18,'div',[[
        'class',
        'lb-outerContainer transition'
      ]
    ],(null as any),(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n      '])),
    (l()(),import0.ɵeld(0,[
      [
        2,
        0
      ]
      ,
      [
        'container',
        1
      ]

    ]
      ,(null as any),15,'div',[[
        'class',
        'lb-container'
      ]
    ],(null as any),(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n        '])),
    (l()(),import0.ɵeld(0,[
      [
        7,
        0
      ]
      ,
      [
        'image',
        1
      ]

    ]
      ,(null as any),0,'img',[[
        'class',
        'lb-image animation fadeIn'
      ]
    ],[
      [
        8,
        'src',
        4
      ]
      ,
      [
        8,
        'hidden',
        0
      ]

    ]
    ,(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n        '])),
    (l()(),import0.ɵeld(0,[
      [
        5,
        0
      ]
      ,
      [
        'navArrow',
        1
      ]

    ]
      ,(null as any),5,'div',[[
        'class',
        'lb-nav'
      ]
      ],[[
        8,
        'hidden',
        0
      ]
    ],(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n          '])),
    (l()(),import0.ɵeld(0,[
      [
        3,
        0
      ]
      ,
      [
        'leftArrow',
        1
      ]

    ]
      ,(null as any),0,'a',[[
        'class',
        'lb-prev'
      ]
      ],[[
        8,
        'hidden',
        0
      ]
      ],[[
        (null as any),
        'click'
      ]
    ],(v,en,$event) => {
      var ad:boolean = true;
      var co:import1.LightboxComponent = v.component;
      if (('click' === en)) {
        const pd_0:any = ((<any>co.prevImage()) !== false);
        ad = (pd_0 && ad);
      }
      return ad;
    },(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n          '])),
    (l()(),import0.ɵeld(0,[
      [
        4,
        0
      ]
      ,
      [
        'rightArrow',
        1
      ]

    ]
      ,(null as any),0,'a',[[
        'class',
        'lb-next'
      ]
      ],[[
        8,
        'hidden',
        0
      ]
      ],[[
        (null as any),
        'click'
      ]
    ],(v,en,$event) => {
      var ad:boolean = true;
      var co:import1.LightboxComponent = v.component;
      if (('click' === en)) {
        const pd_0:any = ((<any>co.nextImage()) !== false);
        ad = (pd_0 && ad);
      }
      return ad;
    },(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n        '])),
    (l()(),import0.ɵted((null as any),['\n        '])),
      (l()(),import0.ɵeld(0,(null as any),(null as any),3,'div',[[
        'class',
        'lb-loader'
      ]
      ],[[
        8,
        'hidden',
        0
      ]
      ],[[
        (null as any),
        'click'
      ]
    ],(v,en,$event) => {
      var ad:boolean = true;
      var co:import1.LightboxComponent = v.component;
      if (('click' === en)) {
        const pd_0:any = ((<any>co.close($event)) !== false);
        ad = (pd_0 && ad);
      }
      return ad;
    },(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n          '])),
      (l()(),import0.ɵeld(0,(null as any),(null as any),0,'a',[[
        'class',
        'lb-cancel'
      ]
    ],(null as any),(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n        '])),
    (l()(),import0.ɵted((null as any),['\n      '])),
    (l()(),import0.ɵted((null as any),['\n    '])),
    (l()(),import0.ɵted((null as any),['\n    '])),
    (l()(),import0.ɵeld(0,[
      [
        6,
        0
      ]
      ,
      [
        'dataContainer',
        1
      ]

    ]
      ,(null as any),18,'div',[[
        'class',
        'lb-dataContainer'
      ]
      ],[[
        8,
        'hidden',
        0
      ]
    ],(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n      '])),
      (l()(),import0.ɵeld(0,(null as any),(null as any),15,'div',[[
        'class',
        'lb-data'
      ]
    ],(null as any),(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n        '])),
      (l()(),import0.ɵeld(0,(null as any),(null as any),7,'div',[[
        'class',
        'lb-details'
      ]
    ],(null as any),(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n          '])),
    (l()(),import0.ɵeld(0,[
      [
        8,
        0
      ]
      ,
      [
        'caption',
        1
      ]

    ]
      ,(null as any),1,'span',[[
        'class',
        'lb-caption animation fadeIn'
      ]
      ],[[
        8,
        'hidden',
        0
      ]
    ],(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),[
      '',
      ''
    ]
    )),
    (l()(),import0.ɵted((null as any),['\n          '])),
    (l()(),import0.ɵeld(0,[
      [
        9,
        0
      ]
      ,
      [
        'number',
        1
      ]

    ]
      ,(null as any),1,'span',[[
        'class',
        'lb-number animation fadeIn'
      ]
      ],[[
        8,
        'hidden',
        0
      ]
    ],(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),[
      '',
      ''
    ]
    )),
    (l()(),import0.ɵted((null as any),['\n        '])),
    (l()(),import0.ɵted((null as any),['\n        '])),
      (l()(),import0.ɵeld(0,(null as any),(null as any),3,'div',[[
        'class',
        'lb-closeContainer'
      ]
    ],(null as any),(null as any),(null as any),(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n          '])),
      (l()(),import0.ɵeld(0,(null as any),(null as any),0,'a',[[
        'class',
        'lb-close'
      ]
      ],(null as any),[[
        (null as any),
        'click'
      ]
    ],(v,en,$event) => {
      var ad:boolean = true;
      var co:import1.LightboxComponent = v.component;
      if (('click' === en)) {
        const pd_0:any = ((<any>co.close($event)) !== false);
        ad = (pd_0 && ad);
      }
      return ad;
    },(null as any),(null as any))),
    (l()(),import0.ɵted((null as any),['\n        '])),
    (l()(),import0.ɵted((null as any),['\n      '])),
    (l()(),import0.ɵted((null as any),['\n    ']))
  ]
  ,(null as any),(ck,v) => {
    var co:import1.LightboxComponent = v.component;
    const currVal_0:any = co.album[co.currentImageIndex].src;
    const currVal_1:any = co.ui.showReloader;
    ck(v,14,0,currVal_0,currVal_1);
    const currVal_2:boolean = !co.ui.showArrowNav;
    ck(v,16,0,currVal_2);
    const currVal_3:boolean = !co.ui.showLeftArrow;
    ck(v,18,0,currVal_3);
    const currVal_4:boolean = !co.ui.showRightArrow;
    ck(v,20,0,currVal_4);
    const currVal_5:boolean = !co.ui.showReloader;
    ck(v,23,0,currVal_5);
    const currVal_6:any = co.ui.showReloader;
    ck(v,30,0,currVal_6);
    const currVal_7:boolean = !co.ui.showCaption;
    ck(v,36,0,currVal_7);
    const currVal_8:any = co.album[co.currentImageIndex].caption;
    ck(v,37,0,currVal_8);
    const currVal_9:boolean = !co.ui.showPageNumber;
    ck(v,39,0,currVal_9);
    const currVal_10:any = co.content.pageNumber;
    ck(v,40,0,currVal_10);
  });
}
function View_LightboxComponent_Host_0(l:any):import0.ɵViewDefinition {
  return import0.ɵvid(0,[
      (l()(),import0.ɵeld(0,(null as any),(null as any),1,'div',[[
        'lb-content',
        ''
      ]
      ],[[
        8,
        'className',
        0
      ]
      ],[[
        (null as any),
        'click'
      ]
    ],(v,en,$event) => {
      var ad:boolean = true;
      if (('click' === en)) {
        const pd_0:any = ((<any>import0.ɵnov(v,1).close($event)) !== false);
        ad = (pd_0 && ad);
      }
      return ad;
    },View_LightboxComponent_0,RenderType_LightboxComponent)),
    import0.ɵdid(2187264,(null as any),0,import1.LightboxComponent,[
      import0.ElementRef,
      import0.Renderer,
      import2.LightboxEvent,
      import0.ElementRef,
      import2.LightboxWindowRef,
      import3.DOCUMENT
    ]
    ,(null as any),(null as any))
  ]
  ,(null as any),(ck,v) => {
    const currVal_0:any = import0.ɵnov(v,1).ui.classList;
    ck(v,0,0,currVal_0);
  });
}
export const LightboxComponentNgFactory:import0.ComponentFactory<import1.LightboxComponent> = import0.ɵccf('[lb-content]',import1.LightboxComponent,View_LightboxComponent_Host_0,{
  album: 'album',
  currentImageIndex: 'currentImageIndex',
  options: 'options',
  cmpRef: 'cmpRef'
}
,{},([] as any[]));
