// angulartics2/uiroutermodule/esm2015/uirouter.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Transition, TransitionService } from '@uirouter/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@uirouter/core";
/**
 * Track Route changes for applications using UI-Router
 *
 * @link https://ui-router.github.io/ng2/docs/latest/
 *
 * referenced: https://github.com/ui-router/sample-app-angular/blob/9adb533b85c0f0fccef23968489cca0a5ec84654/src/app/util/ga.ts
 */
let UIRouterTracking = class UIRouterTracking {
    constructor(transitionService) {
        this.transitionService = transitionService;
    }
    path(trans) {
        return trans.$to().url.format(trans.params());
    }
    trackLocation(settings) {
        const subject = new Subject();
        this.transitionService.onSuccess({}, trans => {
            return subject.next({ url: this.path(trans) });
        }, {
            priority: -10000,
        });
        return subject;
    }
    prepareExternalUrl(url) {
        return url;
    }
};
UIRouterTracking.ctorParameters = () => [
    { type: TransitionService }
];
UIRouterTracking.ɵprov = i0.ɵɵdefineInjectable({ factory: function UIRouterTracking_Factory() { return new UIRouterTracking(i0.ɵɵinject(i1.TransitionService)); }, token: UIRouterTracking, providedIn: "root" });
UIRouterTracking = __decorate([
    Injectable({ providedIn: 'root' })
], UIRouterTracking);
export { UIRouterTracking };
