// angulartics2/esm2015/angulartics2On.js
import { __decorate } from "tslib";
import { AfterContentInit, Directive, ElementRef, Input, NgModule, Renderer2, } from '@angular/core';
import { Angulartics2 } from './angulartics2-core';
let Angulartics2On = class Angulartics2On {
    constructor(elRef, angulartics2, renderer) {
        this.elRef = elRef;
        this.angulartics2 = angulartics2;
        this.renderer = renderer;
        this.angularticsProperties = {};
    }
    ngAfterContentInit() {
        this.renderer.listen(this.elRef.nativeElement, this.angulartics2On || 'click', (event) => this.eventTrack(event));
    }
    eventTrack(event) {
        const action = this.angularticsAction; // || this.inferEventName();
        const properties = Object.assign(Object.assign({}, this.angularticsProperties), { eventType: event.type });
        if (this.angularticsCategory) {
            properties.category = this.angularticsCategory;
        }
        if (this.angularticsLabel) {
            properties.label = this.angularticsLabel;
        }
        if (this.angularticsValue) {
            properties.value = this.angularticsValue;
        }
        this.angulartics2.eventTrack.next({
            action,
            properties,
        });
    }
};
Angulartics2On.ctorParameters = () => [
    { type: ElementRef },
    { type: Angulartics2 },
    { type: Renderer2 }
];
__decorate([
    Input('angulartics2On')
], Angulartics2On.prototype, "angulartics2On", void 0);
__decorate([
    Input()
], Angulartics2On.prototype, "angularticsAction", void 0);
__decorate([
    Input()
], Angulartics2On.prototype, "angularticsCategory", void 0);
__decorate([
    Input()
], Angulartics2On.prototype, "angularticsLabel", void 0);
__decorate([
    Input()
], Angulartics2On.prototype, "angularticsValue", void 0);
__decorate([
    Input()
], Angulartics2On.prototype, "angularticsProperties", void 0);
Angulartics2On = __decorate([
    Directive({ selector: '[angulartics2On]' })
], Angulartics2On);
export { Angulartics2On };
let Angulartics2OnModule = class Angulartics2OnModule {
};
Angulartics2OnModule = __decorate([
    NgModule({
        declarations: [Angulartics2On],
        exports: [Angulartics2On],
    })
], Angulartics2OnModule);
export { Angulartics2OnModule };
