// angulartics2/esm2015/angulartics2.module.js
var Angulartics2Module_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { AngularRouterTracking } from './angular-router';
import { Angulartics2 } from './angulartics2-core';
import { ANGULARTICS2_TOKEN } from './angulartics2-token';
import { Angulartics2On, Angulartics2OnModule } from './angulartics2On';
import { RouterlessTracking } from './routerless';
let Angulartics2Module = Angulartics2Module_1 = class Angulartics2Module {
    static forRoot(settings = {}) {
        return {
            ngModule: Angulartics2Module_1,
            providers: [
                { provide: ANGULARTICS2_TOKEN, useValue: { settings } },
                { provide: RouterlessTracking, useClass: AngularRouterTracking },
                Angulartics2,
            ],
        };
    }
};
Angulartics2Module = Angulartics2Module_1 = __decorate([
    NgModule({
        imports: [Angulartics2OnModule],
        exports: [Angulartics2On],
    })
], Angulartics2Module);
export { Angulartics2Module };
