// angular-2-dropdown-multiselect/bundles/angular-2-dropdown-multiselect.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('rxjs'), require('rxjs/operators'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-2-dropdown-multiselect', ['exports', '@angular/core', '@angular/forms', 'rxjs', 'rxjs/operators', '@angular/common'], factory) :
    (factory((global['angular-2-dropdown-multiselect'] = {}),global.ng.core,global.ng.forms,global.rxjs,global.rxjs.operators,global.ng.common));
}(this, (function (exports,core,forms,rxjs,operators,common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var MultiSelectSearchFilter = (function () {
        function MultiSelectSearchFilter() {
            this._searchCache = {};
            this._searchCacheInclusive = {};
            this._prevSkippedItems = {};
        }
        /**
         * @param {?} options
         * @param {?=} str
         * @param {?=} limit
         * @param {?=} renderLimit
         * @param {?=} searchFunction
         * @return {?}
         */
        MultiSelectSearchFilter.prototype.transform = /**
         * @param {?} options
         * @param {?=} str
         * @param {?=} limit
         * @param {?=} renderLimit
         * @param {?=} searchFunction
         * @return {?}
         */
            function (options, str, limit, renderLimit, searchFunction) {
                if (str === void 0) {
                    str = '';
                }
                if (limit === void 0) {
                    limit = 0;
                }
                if (renderLimit === void 0) {
                    renderLimit = 0;
                }
                str = str.toLowerCase();
                // Drop cache because options were updated
                if (options !== this._lastOptions) {
                    this._lastOptions = options;
                    this._searchCache = {};
                    this._searchCacheInclusive = {};
                    this._prevSkippedItems = {};
                }
                var /** @type {?} */ filteredOpts = this._searchCache.hasOwnProperty(str)
                    ? this._searchCache[str]
                    : this._doSearch(options, str, limit, searchFunction);
                var /** @type {?} */ isUnderLimit = options.length <= limit;
                return isUnderLimit
                    ? filteredOpts
                    : this._limitRenderedItems(filteredOpts, renderLimit);
            };
        /**
         * @param {?} options
         * @param {?} prevOptions
         * @param {?} prevSearchStr
         * @return {?}
         */
        MultiSelectSearchFilter.prototype._getSubsetOptions = /**
         * @param {?} options
         * @param {?} prevOptions
         * @param {?} prevSearchStr
         * @return {?}
         */
            function (options, prevOptions, prevSearchStr) {
                var /** @type {?} */ prevInclusiveOrIdx = this._searchCacheInclusive[prevSearchStr];
                if (prevInclusiveOrIdx === true) {
                    // If have previous results and it was inclusive, do only subsearch
                    return prevOptions;
                }
                else if (typeof prevInclusiveOrIdx === 'number') {
                    // Or reuse prev results with unchecked ones
                    return __spread(prevOptions, options.slice(prevInclusiveOrIdx));
                }
                return options;
            };
        /**
         * @param {?} options
         * @param {?} str
         * @param {?} limit
         * @param {?} searchFunction
         * @return {?}
         */
        MultiSelectSearchFilter.prototype._doSearch = /**
         * @param {?} options
         * @param {?} str
         * @param {?} limit
         * @param {?} searchFunction
         * @return {?}
         */
            function (options, str, limit, searchFunction) {
                var /** @type {?} */ prevStr = str.slice(0, -1);
                var /** @type {?} */ prevResults = this._searchCache[prevStr];
                var /** @type {?} */ prevResultShift = this._prevSkippedItems[prevStr] || 0;
                if (prevResults) {
                    options = this._getSubsetOptions(options, prevResults, prevStr);
                }
                var /** @type {?} */ optsLength = options.length;
                var /** @type {?} */ maxFound = limit > 0 ? Math.min(limit, optsLength) : optsLength;
                var /** @type {?} */ regexp = searchFunction(str);
                var /** @type {?} */ filteredOpts = [];
                var /** @type {?} */ i = 0, /** @type {?} */ founded = 0, /** @type {?} */ removedFromPrevResult = 0;
                var /** @type {?} */ doesOptionMatch = function (option) { return regexp.test(option.name); };
                var /** @type {?} */ getChildren = function (option) {
                    return options.filter(function (child) { return child.parentId === option.id; });
                };
                var /** @type {?} */ getParent = function (option) {
                    return options.find(function (parent) { return option.parentId === parent.id; });
                };
                var /** @type {?} */ foundFn = function (item) { filteredOpts.push(item); founded++; };
                var /** @type {?} */ notFoundFn = prevResults ? function () { return removedFromPrevResult++; } : function () { };
                for (; i < optsLength && founded < maxFound; ++i) {
                    var /** @type {?} */ option = options[i];
                    var /** @type {?} */ directMatch = doesOptionMatch(option);
                    if (directMatch) {
                        foundFn(option);
                        continue;
                    }
                    if (typeof option.parentId === 'undefined') {
                        var /** @type {?} */ childrenMatch = getChildren(option).some(doesOptionMatch);
                        if (childrenMatch) {
                            foundFn(option);
                            continue;
                        }
                    }
                    if (typeof option.parentId !== 'undefined') {
                        var /** @type {?} */ parentMatch = doesOptionMatch(getParent(option));
                        if (parentMatch) {
                            foundFn(option);
                            continue;
                        }
                    }
                    notFoundFn();
                }
                var /** @type {?} */ totalIterations = i + prevResultShift;
                this._searchCache[str] = filteredOpts;
                this._searchCacheInclusive[str] = i === optsLength || totalIterations;
                this._prevSkippedItems[str] = removedFromPrevResult + prevResultShift;
                return filteredOpts;
            };
        /**
         * @template T
         * @param {?} items
         * @param {?} limit
         * @return {?}
         */
        MultiSelectSearchFilter.prototype._limitRenderedItems = /**
         * @template T
         * @param {?} items
         * @param {?} limit
         * @return {?}
         */
            function (items, limit) {
                return items.length > limit && limit > 0 ? items.slice(0, limit) : items;
            };
        MultiSelectSearchFilter.decorators = [
            { type: core.Pipe, args: [{
                        name: 'searchFilter'
                    },] }
        ];
        return MultiSelectSearchFilter;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var AutofocusDirective = (function () {
        function AutofocusDirective(elemRef) {
            this.elemRef = elemRef;
        }
        Object.defineProperty(AutofocusDirective.prototype, "element", {
            get: /**
             * @return {?}
             */ function () {
                return this.elemRef.nativeElement;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        AutofocusDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.focus();
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        AutofocusDirective.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                var /** @type {?} */ ssAutofocusChange = changes["ssAutofocus"];
                if (ssAutofocusChange && !ssAutofocusChange.isFirstChange()) {
                    this.focus();
                }
            };
        /**
         * @return {?}
         */
        AutofocusDirective.prototype.focus = /**
         * @return {?}
         */
            function () {
                if (this.ssAutofocus) {
                    return;
                }
                this.element.focus && this.element.focus();
            };
        AutofocusDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[ssAutofocus]'
                    },] }
        ];
        /** @nocollapse */
        AutofocusDirective.ctorParameters = function () {
            return [
                { type: core.ElementRef, decorators: [{ type: core.Host },] },
            ];
        };
        AutofocusDirective.propDecorators = {
            "ssAutofocus": [{ type: core.Input },],
        };
        return AutofocusDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /*
     * Angular 2 Dropdown Multiselect for Bootstrap
     *
     * Simon Lindh
     * https://github.com/softsimon/angular-2-dropdown-multiselect
     */
    var /** @type {?} */ MULTISELECT_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return MultiselectDropdownComponent; }),
        multi: true,
    };
    var MultiselectDropdownComponent = (function () {
        function MultiselectDropdownComponent(element, fb, searchFilter, differs, cdRef) {
            this.element = element;
            this.fb = fb;
            this.searchFilter = searchFilter;
            this.cdRef = cdRef;
            this.filterControl = this.fb.control('');
            this.disabled = false;
            this.disabledSelection = false;
            this.searchFunction = this._escapeRegExp;
            this.selectionLimitReached = new core.EventEmitter();
            this.dropdownClosed = new core.EventEmitter();
            this.dropdownOpened = new core.EventEmitter();
            this.onAdded = new core.EventEmitter();
            this.onRemoved = new core.EventEmitter();
            this.onLazyLoad = new core.EventEmitter();
            this.onFilter = this.filterControl.valueChanges;
            this.destroyed$ = new rxjs.Subject();
            this.filteredOptions = [];
            this.lazyLoadOptions = [];
            this.renderFilteredOptions = [];
            this.model = [];
            this.prevModel = [];
            this.numSelected = 0;
            this.renderItems = true;
            this.checkAllSearchRegister = new Set();
            this.checkAllStatus = false;
            this.loadedValueIds = [];
            this._focusBack = false;
            this.defaultSettings = {
                closeOnClickOutside: true,
                pullRight: false,
                enableSearch: false,
                searchRenderLimit: 0,
                searchRenderAfter: 1,
                searchMaxLimit: 0,
                searchMaxRenderedItems: 0,
                checkedStyle: 'checkboxes',
                buttonClasses: 'btn btn-primary dropdown-toggle',
                containerClasses: 'dropdown-inline',
                selectionLimit: 0,
                minSelectionLimit: 0,
                closeOnSelect: false,
                autoUnselect: false,
                showCheckAll: false,
                showUncheckAll: false,
                fixedTitle: false,
                dynamicTitleMaxItems: 3,
                maxHeight: '300px',
                isLazyLoad: false,
                stopScrollPropagation: false,
                loadViewDistance: 1,
                selectAddedValues: false,
                ignoreLabels: false,
                maintainSelectionOrderInTitle: false,
                focusBack: true
            };
            this.defaultTexts = {
                checkAll: 'Check all',
                uncheckAll: 'Uncheck all',
                checked: 'checked',
                checkedPlural: 'checked',
                searchPlaceholder: 'Search...',
                searchEmptyResult: 'Nothing found...',
                searchNoRenderText: 'Type in search box to see results...',
                defaultTitle: 'Select',
                allSelected: 'All selected',
            };
            this._isVisible = false;
            this._workerDocClicked = false;
            this.onModelChange = function (_) { };
            this.onModelTouched = function () { };
            this.differ = differs.find([]).create(null);
            this.settings = this.defaultSettings;
            this.texts = this.defaultTexts;
        }
        Object.defineProperty(MultiselectDropdownComponent.prototype, "focusBack", {
            get: /**
             * @return {?}
             */ function () {
                return this.settings.focusBack && this._focusBack;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.clickedOutside = /**
         * @return {?}
         */
            function () {
                if (!this.isVisible || !this.settings.closeOnClickOutside) {
                    return;
                }
                this.isVisible = false;
                this._focusBack = true;
                this.dropdownClosed.emit();
            };
        Object.defineProperty(MultiselectDropdownComponent.prototype, "isVisible", {
            get: /**
             * @return {?}
             */ function () {
                return this._isVisible;
            },
            set: /**
             * @param {?} val
             * @return {?}
             */ function (val) {
                this._isVisible = val;
                this._workerDocClicked = val ? false : this._workerDocClicked;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MultiselectDropdownComponent.prototype, "searchLimit", {
            get: /**
             * @return {?}
             */ function () {
                return this.settings.searchRenderLimit;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MultiselectDropdownComponent.prototype, "searchRenderAfter", {
            get: /**
             * @return {?}
             */ function () {
                return this.settings.searchRenderAfter;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MultiselectDropdownComponent.prototype, "searchLimitApplied", {
            get: /**
             * @return {?}
             */ function () {
                return this.searchLimit > 0 && this.options.length > this.searchLimit;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} option
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.getItemStyle = /**
         * @param {?} option
         * @return {?}
         */
            function (option) {
                if (!option.isLabel) ;
                if (option.disabled) ;
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.getItemStyleSelectionDisabled = /**
         * @return {?}
         */
            function () {
                if (this.disabledSelection) {
                    return { cursor: 'default' };
                }
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.title = this.texts.defaultTitle || '';
                this.filterControl.valueChanges.pipe(operators.takeUntil(this.destroyed$)).subscribe(function () {
                    _this.updateRenderItems();
                    if (_this.settings.isLazyLoad) {
                        _this.load();
                    }
                });
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                var _this = this;
                if (changes['options']) {
                    this.options = this.options || [];
                    this.parents = this.options
                        .filter(function (option) { return typeof option.parentId === 'number'; })
                        .map(function (option) { return option.parentId; });
                    this.updateRenderItems();
                    if (this.settings.isLazyLoad &&
                        this.settings.selectAddedValues &&
                        this.loadedValueIds.length === 0) {
                        this.loadedValueIds = this.loadedValueIds.concat(changes["options"].currentValue.map(function (value) { return value.id; }));
                    }
                    if (this.settings.isLazyLoad &&
                        this.settings.selectAddedValues &&
                        changes["options"].previousValue) {
                        var /** @type {?} */ addedValues_1 = changes["options"].currentValue.filter(function (value) { return _this.loadedValueIds.indexOf(value.id) === -1; });
                        this.loadedValueIds.concat(addedValues_1.map(function (value) { return value.id; }));
                        if (this.checkAllStatus) {
                            this.addChecks(addedValues_1);
                        }
                        else if (this.checkAllSearchRegister.size > 0) {
                            this.checkAllSearchRegister.forEach(function (searchValue) {
                                return _this.addChecks(_this.applyFilters(addedValues_1, searchValue));
                            });
                        }
                    }
                    if (this.texts) {
                        this.updateTitle();
                    }
                    this.fireModelChange();
                }
                if (changes['settings']) {
                    this.settings = __assign({}, this.defaultSettings, this.settings);
                }
                if (changes['texts']) {
                    this.texts = __assign({}, this.defaultTexts, this.texts);
                    if (!changes['texts'].isFirstChange()) {
                        this.updateTitle();
                    }
                }
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.destroyed$.next();
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.updateRenderItems = /**
         * @return {?}
         */
            function () {
                this.renderItems =
                    !this.searchLimitApplied ||
                        this.filterControl.value.length >= this.searchRenderAfter;
                this.filteredOptions = this.applyFilters(this.options, this.settings.isLazyLoad ? '' : this.filterControl.value);
                this.renderFilteredOptions = this.renderItems ? this.filteredOptions : [];
                this.focusedItem = undefined;
            };
        /**
         * @param {?} options
         * @param {?} value
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.applyFilters = /**
         * @param {?} options
         * @param {?} value
         * @return {?}
         */
            function (options, value) {
                return this.searchFilter.transform(options, value, this.settings.searchMaxLimit, this.settings.searchMaxRenderedItems, this.searchFunction);
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.fireModelChange = /**
         * @return {?}
         */
            function () {
                if (this.model != this.prevModel) {
                    this.prevModel = this.model;
                    this.onModelChange(this.model);
                    this.onModelTouched();
                    this.cdRef.markForCheck();
                }
            };
        /**
         * @param {?} value
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.writeValue = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                if (value !== undefined && value !== null) {
                    this.model = Array.isArray(value) ? value : [value];
                    this.ngDoCheck();
                }
                else {
                    this.model = [];
                }
            };
        /**
         * @param {?} fn
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.registerOnChange = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this.onModelChange = fn;
            };
        /**
         * @param {?} fn
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.registerOnTouched = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this.onModelTouched = fn;
            };
        /**
         * @param {?} isDisabled
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.setDisabledState = /**
         * @param {?} isDisabled
         * @return {?}
         */
            function (isDisabled) {
                this.disabled = isDisabled;
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.ngDoCheck = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ changes = this.differ.diff(this.model);
                if (changes) {
                    this.updateNumSelected();
                    this.updateTitle();
                }
            };
        /**
         * @param {?} _c
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.validate = /**
         * @param {?} _c
         * @return {?}
         */
            function (_c) {
                var _this = this;
                if (this.model && this.model.length) {
                    return {
                        required: {
                            valid: false
                        }
                    };
                }
                if (this.options.filter(function (o) { return _this.model.indexOf(o.id) && !o.disabled; }).length === 0) {
                    return {
                        selection: {
                            valid: false
                        }
                    };
                }
                return null;
            };
        /**
         * @param {?} _fn
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.registerOnValidatorChange = /**
         * @param {?} _fn
         * @return {?}
         */
            function (_fn) {
                throw new Error('Method not implemented.');
            };
        /**
         * @param {?} event
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.clearSearch = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.maybeStopPropagation(event);
                this.filterControl.setValue('');
            };
        /**
         * @param {?=} e
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.toggleDropdown = /**
         * @param {?=} e
         * @return {?}
         */
            function (e) {
                if (this.isVisible) {
                    this._focusBack = true;
                }
                this.isVisible = !this.isVisible;
                this.isVisible ? this.dropdownOpened.emit() : this.dropdownClosed.emit();
                this.focusedItem = undefined;
            };
        /**
         * @param {?=} e
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.closeDropdown = /**
         * @param {?=} e
         * @return {?}
         */
            function (e) {
                this.isVisible = true;
                this.toggleDropdown(e);
            };
        /**
         * @param {?} option
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.isSelected = /**
         * @param {?} option
         * @return {?}
         */
            function (option) {
                return this.model && this.model.indexOf(option.id) > -1;
            };
        /**
         * @param {?} _event
         * @param {?} option
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.setSelected = /**
         * @param {?} _event
         * @param {?} option
         * @return {?}
         */
            function (_event, option) {
                var _this = this;
                if (option.isLabel) {
                    return;
                }
                if (option.disabled) {
                    return;
                }
                if (this.disabledSelection) {
                    return;
                }
                setTimeout(function () {
                    _this.maybeStopPropagation(_event);
                    _this.maybePreventDefault(_event);
                    var /** @type {?} */ index = _this.model.indexOf(option.id);
                    var /** @type {?} */ isAtSelectionLimit = _this.settings.selectionLimit > 0 &&
                        _this.model.length >= _this.settings.selectionLimit;
                    var /** @type {?} */ removeItem = function (idx, id) {
                        _this.model.splice(idx, 1);
                        _this.onRemoved.emit(id);
                        if (_this.settings.isLazyLoad &&
                            _this.lazyLoadOptions.some(function (val) { return val.id === id; })) {
                            _this.lazyLoadOptions.splice(_this.lazyLoadOptions.indexOf(_this.lazyLoadOptions.find(function (val) { return val.id === id; })), 1);
                        }
                    };
                    if (index > -1) {
                        if (_this.settings.minSelectionLimit === undefined ||
                            _this.numSelected > _this.settings.minSelectionLimit) {
                            removeItem(index, option.id);
                        }
                        var /** @type {?} */ parentIndex = option.parentId && _this.model.indexOf(option.parentId);
                        if (parentIndex > -1) {
                            removeItem(parentIndex, option.parentId);
                        }
                        else if (_this.parents.indexOf(option.id) > -1) {
                            _this.options
                                .filter(function (child) {
                                return _this.model.indexOf(child.id) > -1 &&
                                    child.parentId === option.id;
                            })
                                .forEach(function (child) {
                                return removeItem(_this.model.indexOf(child.id), child.id);
                            });
                        }
                    }
                    else if (isAtSelectionLimit && !_this.settings.autoUnselect) {
                        _this.selectionLimitReached.emit(_this.model.length);
                        return;
                    }
                    else {
                        var /** @type {?} */ addItem_1 = function (id) {
                            _this.model.push(id);
                            _this.onAdded.emit(id);
                            if (_this.settings.isLazyLoad &&
                                !_this.lazyLoadOptions.some(function (val) { return val.id === id; })) {
                                _this.lazyLoadOptions.push(option);
                            }
                        };
                        addItem_1(option.id);
                        if (!isAtSelectionLimit) {
                            if (option.parentId && !_this.settings.ignoreLabels) {
                                var /** @type {?} */ children = _this.options.filter(function (child) {
                                    return child.id !== option.id && child.parentId === option.parentId;
                                });
                                if (children.every(function (child) { return _this.model.indexOf(child.id) > -1; })) {
                                    addItem_1(option.parentId);
                                }
                            }
                            else if (_this.parents.indexOf(option.id) > -1) {
                                var /** @type {?} */ children = _this.options.filter(function (child) {
                                    return _this.model.indexOf(child.id) < 0 && child.parentId === option.id;
                                });
                                children.forEach(function (child) { return addItem_1(child.id); });
                            }
                        }
                        else {
                            removeItem(0, _this.model[0]);
                        }
                    }
                    if (_this.settings.closeOnSelect) {
                        _this.toggleDropdown();
                    }
                    _this.model = _this.model.slice();
                    _this.fireModelChange();
                }, 0);
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.updateNumSelected = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.numSelected =
                    this.model.filter(function (id) { return _this.parents.indexOf(id) < 0; }).length || 0;
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.updateTitle = /**
         * @return {?}
         */
            function () {
                var _this = this;
                var /** @type {?} */ numSelectedOptions = this.options.length;
                if (this.settings.ignoreLabels) {
                    numSelectedOptions = this.options.filter(function (option) { return !option.isLabel; }).length;
                }
                if (this.numSelected === 0 || this.settings.fixedTitle) {
                    this.title = this.texts ? this.texts.defaultTitle : '';
                }
                else if (this.settings.displayAllSelectedText &&
                    this.model.length === numSelectedOptions) {
                    this.title = this.texts ? this.texts.allSelected : '';
                }
                else if (this.settings.dynamicTitleMaxItems &&
                    this.settings.dynamicTitleMaxItems >= this.numSelected) {
                    var /** @type {?} */ useOptions_1 = this.settings.isLazyLoad && this.lazyLoadOptions.length
                        ? this.lazyLoadOptions
                        : this.options;
                    var /** @type {?} */ titleSelections = void 0;
                    if (this.settings.maintainSelectionOrderInTitle) {
                        var /** @type {?} */ optionIds_1 = useOptions_1.map(function (selectOption, idx) { return selectOption.id; });
                        titleSelections = this.model
                            .map(function (selectedId) { return optionIds_1.indexOf(selectedId); })
                            .filter(function (optionIndex) { return optionIndex > -1; })
                            .map(function (optionIndex) { return useOptions_1[optionIndex]; });
                    }
                    else {
                        titleSelections = useOptions_1.filter(function (option) { return _this.model.indexOf(option.id) > -1; });
                    }
                    this.title = titleSelections.map(function (option) { return option.name; }).join(', ');
                }
                else {
                    this.title =
                        this.numSelected +
                            ' ' +
                            (this.numSelected === 1
                                ? this.texts.checked
                                : this.texts.checkedPlural);
                }
                this.cdRef.markForCheck();
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.searchFilterApplied = /**
         * @return {?}
         */
            function () {
                return (this.settings.enableSearch &&
                    this.filterControl.value &&
                    this.filterControl.value.length > 0);
            };
        /**
         * @param {?} options
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.addChecks = /**
         * @param {?} options
         * @return {?}
         */
            function (options) {
                var _this = this;
                var /** @type {?} */ checkedOptions = options
                    .filter(function (option) {
                    if (!option.disabled &&
                        (_this.model.indexOf(option.id) === -1 &&
                            !(_this.settings.ignoreLabels && option.isLabel))) {
                        _this.onAdded.emit(option.id);
                        return true;
                    }
                    return false;
                })
                    .map(function (option) { return option.id; });
                this.model = this.model.concat(checkedOptions);
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.checkAll = /**
         * @return {?}
         */
            function () {
                if (!this.disabledSelection) {
                    this.addChecks(!this.searchFilterApplied() ? this.options : this.filteredOptions);
                    if (this.settings.isLazyLoad && this.settings.selectAddedValues) {
                        if (this.searchFilterApplied() && !this.checkAllStatus) {
                            this.checkAllSearchRegister.add(this.filterControl.value);
                        }
                        else {
                            this.checkAllSearchRegister.clear();
                            this.checkAllStatus = true;
                        }
                        this.load();
                    }
                    this.fireModelChange();
                }
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.uncheckAll = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (!this.disabledSelection) {
                    var /** @type {?} */ checkedOptions = this.model;
                    var /** @type {?} */ unCheckedOptions_1 = !this.searchFilterApplied()
                        ? this.model
                        : this.filteredOptions.map(function (option) { return option.id; });
                    // set unchecked options only to the ones that were checked
                    // set unchecked options only to the ones that were checked
                    unCheckedOptions_1 = checkedOptions.filter(function (item) { return unCheckedOptions_1.indexOf(item) > -1; });
                    this.model = this.model.filter(function (id) {
                        if ((unCheckedOptions_1.indexOf(id) < 0 &&
                            _this.settings.minSelectionLimit === undefined) ||
                            unCheckedOptions_1.indexOf(id) < _this.settings.minSelectionLimit) {
                            return true;
                        }
                        else {
                            _this.onRemoved.emit(id);
                            return false;
                        }
                    });
                    if (this.settings.isLazyLoad && this.settings.selectAddedValues) {
                        if (this.searchFilterApplied()) {
                            if (this.checkAllSearchRegister.has(this.filterControl.value)) {
                                this.checkAllSearchRegister.delete(this.filterControl.value);
                                this.checkAllSearchRegister.forEach(function (searchTerm) {
                                    var /** @type {?} */ filterOptions = this.applyFilters(this.options.filter(function (option) { return unCheckedOptions_1.indexOf(option.id) > -1; }), searchTerm);
                                    this.addChecks(filterOptions);
                                });
                            }
                        }
                        else {
                            this.checkAllSearchRegister.clear();
                            this.checkAllStatus = false;
                        }
                        this.load();
                    }
                    this.fireModelChange();
                }
            };
        /**
         * @param {?} event
         * @param {?} option
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.preventCheckboxCheck = /**
         * @param {?} event
         * @param {?} option
         * @return {?}
         */
            function (event, option) {
                if (option.disabled ||
                    (this.settings.selectionLimit &&
                        !this.settings.autoUnselect &&
                        this.model.length >= this.settings.selectionLimit &&
                        this.model.indexOf(option.id) === -1 &&
                        this.maybePreventDefault(event))) {
                    this.maybePreventDefault(event);
                }
            };
        /**
         * @param {?=} option
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.isCheckboxDisabled = /**
         * @param {?=} option
         * @return {?}
         */
            function (option) {
                return this.disabledSelection || option && option.disabled;
            };
        /**
         * @param {?} ev
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.checkScrollPosition = /**
         * @param {?} ev
         * @return {?}
         */
            function (ev) {
                var /** @type {?} */ scrollTop = ev.target.scrollTop;
                var /** @type {?} */ scrollHeight = ev.target.scrollHeight;
                var /** @type {?} */ scrollElementHeight = ev.target.clientHeight;
                var /** @type {?} */ roundingPixel = 1;
                var /** @type {?} */ gutterPixel = 1;
                if (scrollTop >=
                    scrollHeight -
                        (1 + this.settings.loadViewDistance) * scrollElementHeight -
                        roundingPixel -
                        gutterPixel) {
                    this.load();
                }
            };
        /**
         * @param {?} ev
         * @param {?} element
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.checkScrollPropagation = /**
         * @param {?} ev
         * @param {?} element
         * @return {?}
         */
            function (ev, element) {
                var /** @type {?} */ scrollTop = element.scrollTop;
                var /** @type {?} */ scrollHeight = element.scrollHeight;
                var /** @type {?} */ scrollElementHeight = element.clientHeight;
                if ((ev.deltaY > 0 && scrollTop + scrollElementHeight >= scrollHeight) ||
                    (ev.deltaY < 0 && scrollTop <= 0)) {
                    ev = ev || window.event;
                    this.maybePreventDefault(ev);
                    ev.returnValue = false;
                }
            };
        /**
         * @param {?} idx
         * @param {?} selectOption
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.trackById = /**
         * @param {?} idx
         * @param {?} selectOption
         * @return {?}
         */
            function (idx, selectOption) {
                return selectOption.id;
            };
        /**
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.load = /**
         * @return {?}
         */
            function () {
                this.onLazyLoad.emit({
                    length: this.options.length,
                    filter: this.filterControl.value,
                    checkAllSearches: this.checkAllSearchRegister,
                    checkAllStatus: this.checkAllStatus,
                });
            };
        /**
         * @param {?} dir
         * @param {?=} e
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.focusItem = /**
         * @param {?} dir
         * @param {?=} e
         * @return {?}
         */
            function (dir, e) {
                if (!this.isVisible) {
                    return;
                }
                this.maybePreventDefault(e);
                var /** @type {?} */ idx = this.filteredOptions.indexOf(this.focusedItem);
                if (idx === -1) {
                    this.focusedItem = this.filteredOptions[0];
                    return;
                }
                var /** @type {?} */ nextIdx = idx + dir;
                var /** @type {?} */ newIdx = nextIdx < 0
                    ? this.filteredOptions.length - 1
                    : nextIdx % this.filteredOptions.length;
                this.focusedItem = this.filteredOptions[newIdx];
            };
        /**
         * @param {?=} e
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.maybePreventDefault = /**
         * @param {?=} e
         * @return {?}
         */
            function (e) {
                if (e && e.preventDefault) {
                    e.preventDefault();
                }
            };
        /**
         * @param {?=} e
         * @return {?}
         */
        MultiselectDropdownComponent.prototype.maybeStopPropagation = /**
         * @param {?=} e
         * @return {?}
         */
            function (e) {
                if (e && e.stopPropagation) {
                    e.stopPropagation();
                }
            };
        /**
         * @param {?} str
         * @return {?}
         */
        MultiselectDropdownComponent.prototype._escapeRegExp = /**
         * @param {?} str
         * @return {?}
         */
            function (str) {
                var /** @type {?} */ regExpStr = str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
                return new RegExp(regExpStr, 'i');
            };
        MultiselectDropdownComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ss-multiselect-dropdown',
                        template: "<div class=\"dropdown\" [ngClass]=\"settings.containerClasses\" [class.open]=\"isVisible\" (offClick)=\"clickedOutside()\">\n  <button type=\"button\" class=\"dropdown-toggle\" [ngClass]=\"settings.buttonClasses\" (click)=\"toggleDropdown($event)\" [disabled]=\"disabled\"\n    [ssAutofocus]=\"!focusBack\">\n    {{ title }}\n    <span class=\"caret\"></span>\n  </button>\n  <div #scroller *ngIf=\"isVisible\" class=\"dropdown-menu\" [ngClass]=\"{'chunkydropdown-menu': settings.checkedStyle == 'visual' }\"\n    (scroll)=\"settings.isLazyLoad ? checkScrollPosition($event) : null\" (wheel)=\"settings.stopScrollPropagation ? checkScrollPropagation($event, scroller) : null\"\n    [class.pull-right]=\"settings.pullRight\" [class.dropdown-menu-right]=\"settings.pullRight\" [style.max-height]=\"settings.maxHeight\"\n    style=\"display: block; height: auto; overflow-y: auto;\" (keydown.tab)=\"focusItem(1, $event)\" (keydown.shift.tab)=\"focusItem(-1, $event)\">\n    <div class=\"input-group search-container\" *ngIf=\"settings.enableSearch\">\n      <div class=\"input-group-prepend\">\n        <span class=\"input-group-text\" id=\"basic-addon1\">\n          <i class=\"fa fa-search\" aria-hidden=\"true\"></i>\n        </span>\n      </div>\n      <input type=\"text\" class=\"form-control\" ssAutofocus [formControl]=\"filterControl\" [placeholder]=\"texts.searchPlaceholder\"\n        class=\"form-control\">\n      <div class=\"input-group-append\" *ngIf=\"filterControl.value.length>0\">\n        <button class=\"btn btn-default btn-secondary\" type=\"button\" (click)=\"clearSearch($event)\">\n          <i class=\"fa fa-times\"></i>\n        </button>\n      </div>\n    </div>\n    <a role=\"menuitem\" href=\"javascript:;\" tabindex=\"-1\" class=\"dropdown-item check-control check-control-check\" *ngIf=\"settings.showCheckAll && !disabledSelection\"\n      (click)=\"checkAll()\">\n      <span style=\"width: 16px;\"><span [ngClass]=\"{'glyphicon glyphicon-ok': settings.checkedStyle !== 'fontawesome','fa fa-check': settings.checkedStyle === 'fontawesome'}\"></span></span>\n      {{ texts.checkAll }}\n    </a>\n    <a role=\"menuitem\" href=\"javascript:;\" tabindex=\"-1\" class=\"dropdown-item check-control check-control-uncheck\" *ngIf=\"settings.showUncheckAll && !disabledSelection\"\n      (click)=\"uncheckAll()\">\n      <span style=\"width: 16px;\"><span [ngClass]=\"{'glyphicon glyphicon-remove': settings.checkedStyle !== 'fontawesome','fa fa-times': settings.checkedStyle === 'fontawesome'}\"></span></span>\n      {{ texts.uncheckAll }}\n    </a>\n    <a *ngIf=\"settings.showCheckAll || settings.showUncheckAll\" href=\"javascript:;\" class=\"dropdown-divider divider\"></a>\n    <a *ngIf=\"!renderItems\" href=\"javascript:;\" class=\"dropdown-item empty\">{{ texts.searchNoRenderText }}</a>\n    <a *ngIf=\"renderItems && !renderFilteredOptions.length\" href=\"javascript:;\" class=\"dropdown-item empty\">{{ texts.searchEmptyResult }}</a>\n    <a class=\"dropdown-item\" href=\"javascript:;\" *ngFor=\"let option of renderFilteredOptions; trackBy: trackById\" [class.active]=\"isSelected(option)\"\n      [ngStyle]=\"getItemStyle(option)\" [ngClass]=\"option.classes\" [class.dropdown-header]=\"option.isLabel\" [ssAutofocus]=\"option !== focusedItem\"\n      tabindex=\"-1\" (click)=\"setSelected($event, option)\" (keydown.space)=\"setSelected($event, option)\" (keydown.enter)=\"setSelected($event, option)\">\n      <span *ngIf=\"!option.isLabel; else label\" role=\"menuitem\" tabindex=\"-1\" [style.padding-left]=\"this.parents.length>0&&this.parents.indexOf(option.id)<0&&'30px'\"\n        [ngStyle]=\"getItemStyleSelectionDisabled()\">\n        <ng-container [ngSwitch]=\"settings.checkedStyle\">\n          <input *ngSwitchCase=\"'checkboxes'\" type=\"checkbox\" [checked]=\"isSelected(option)\" (click)=\"preventCheckboxCheck($event, option)\"\n            [disabled]=\"isCheckboxDisabled(option)\" [ngStyle]=\"getItemStyleSelectionDisabled()\" />\n          <span *ngSwitchCase=\"'glyphicon'\" style=\"width: 16px;\" class=\"glyphicon\" [class.glyphicon-ok]=\"isSelected(option)\" [class.glyphicon-lock]=\"isCheckboxDisabled(option)\"></span>\n          <span *ngSwitchCase=\"'fontawesome'\" style=\"width: 16px;display: inline-block;\">\n            <span *ngIf=\"isSelected(option)\"><i class=\"fa fa-check\" aria-hidden=\"true\"></i></span>\n            <span *ngIf=\"isCheckboxDisabled(option)\"><i class=\"fa fa-lock\" aria-hidden=\"true\"></i></span>\n          </span>\n          <span *ngSwitchCase=\"'visual'\" style=\"display:block;float:left; border-radius: 0.2em; border: 0.1em solid rgba(44, 44, 44, 0.63);background:rgba(0, 0, 0, 0.1);width: 5.5em;\">\n            <div class=\"slider\" [ngClass]=\"{'slideron': isSelected(option)}\">\n              <img *ngIf=\"option.image != null\" [src]=\"option.image\" style=\"height: 100%; width: 100%; object-fit: contain\" />\n              <div *ngIf=\"option.image == null\" style=\"height: 100%; width: 100%;text-align: center; display: table; background-color:rgba(0, 0, 0, 0.74)\">\n                <div class=\"content_wrapper\">\n                  <span style=\"font-size:3em;color:white\" class=\"glyphicon glyphicon-eye-close\"></span>\n                </div>\n              </div>\n            </div>\n          </span>\n        </ng-container>\n        <span [ngClass]=\"{'chunkyrow': settings.checkedStyle == 'visual' }\" [class.disabled]=\"isCheckboxDisabled(option)\" [ngClass]=\"settings.itemClasses\"\n          [style.font-weight]=\"this.parents.indexOf(option.id)>=0?'bold':'normal'\">\n          {{ option.name }}\n        </span>\n      </span>\n      <ng-template #label>\n        <span [class.disabled]=\"isCheckboxDisabled(option)\">{{ option.name }}</span>\n      </ng-template>\n    </a>\n  </div>\n</div>\n",
                        providers: [MULTISELECT_VALUE_ACCESSOR, MultiSelectSearchFilter],
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        styles: ["a{outline:0!important}.dropdown-inline{display:inline-block}.dropdown-toggle .caret{margin-left:4px;white-space:nowrap;display:inline-block}.chunkydropdown-menu{min-width:20em}.chunkyrow{line-height:2;margin-left:1em;font-size:2em}.slider{width:3.8em;height:3.8em;display:block;transition:all 125ms linear;margin-left:.125em;margin-top:auto}.slideron{margin-left:1.35em}.content_wrapper{display:table-cell;vertical-align:middle}.search-container{padding:0 5px 5px}"]
                    }] }
        ];
        /** @nocollapse */
        MultiselectDropdownComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef, },
                { type: forms.FormBuilder, },
                { type: MultiSelectSearchFilter, },
                { type: core.IterableDiffers, },
                { type: core.ChangeDetectorRef, },
            ];
        };
        MultiselectDropdownComponent.propDecorators = {
            "options": [{ type: core.Input },],
            "settings": [{ type: core.Input },],
            "texts": [{ type: core.Input },],
            "disabled": [{ type: core.Input },],
            "disabledSelection": [{ type: core.Input },],
            "searchFunction": [{ type: core.Input },],
            "selectionLimitReached": [{ type: core.Output },],
            "dropdownClosed": [{ type: core.Output },],
            "dropdownOpened": [{ type: core.Output },],
            "onAdded": [{ type: core.Output },],
            "onRemoved": [{ type: core.Output },],
            "onLazyLoad": [{ type: core.Output },],
            "onFilter": [{ type: core.Output },],
        };
        return MultiselectDropdownComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var OffClickDirective = (function () {
        function OffClickDirective() {
            this.onOffClick = new core.EventEmitter();
        }
        /**
         * @param {?} event
         * @return {?}
         */
        OffClickDirective.prototype.onClick = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this._clickEvent = event;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        OffClickDirective.prototype.onTouch = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this._touchEvent = event;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        OffClickDirective.prototype.onDocumentClick = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (event !== this._clickEvent) {
                    this.onOffClick.emit(event);
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        OffClickDirective.prototype.onDocumentTouch = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (event !== this._touchEvent) {
                    this.onOffClick.emit(event);
                }
            };
        OffClickDirective.decorators = [
            { type: core.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: '[offClick]',
                    },] }
        ];
        /** @nocollapse */
        OffClickDirective.propDecorators = {
            "onOffClick": [{ type: core.Output, args: ['offClick',] },],
            "onClick": [{ type: core.HostListener, args: ['click', ['$event'],] },],
            "onTouch": [{ type: core.HostListener, args: ['touchstart', ['$event'],] },],
            "onDocumentClick": [{ type: core.HostListener, args: ['document:click', ['$event'],] },],
            "onDocumentTouch": [{ type: core.HostListener, args: ['document:touchstart', ['$event'],] },],
        };
        return OffClickDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var MultiselectDropdownModule = (function () {
        function MultiselectDropdownModule() {
        }
        MultiselectDropdownModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [common.CommonModule, forms.ReactiveFormsModule],
                        exports: [
                            MultiselectDropdownComponent,
                            MultiSelectSearchFilter,
                        ],
                        declarations: [
                            MultiselectDropdownComponent,
                            MultiSelectSearchFilter,
                            AutofocusDirective,
                            OffClickDirective
                        ],
                    },] }
        ];
        return MultiselectDropdownModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.MultiSelectSearchFilter = MultiSelectSearchFilter;
    exports.MultiselectDropdownModule = MultiselectDropdownModule;
    exports.MultiselectDropdownComponent = MultiselectDropdownComponent;
    exports.a = AutofocusDirective;
    exports.b = OffClickDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
