// angular2-lightbox/compiled/lightbox-overlay.component.ngfactory.ts
/**
 * @fileoverview This file is generated by the Angular template compiler.
 * Do not edit.
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride}
 */
 /* tslint:disable */


import * as import0 from '@angular/core';
import * as import1 from '../lightbox-overlay.component';
import * as import2 from '../lightbox-event.service';
import * as import3 from '@angular/platform-browser';
const styles_LightboxOverlayComponent:any[] = ([] as any[]);
export const RenderType_LightboxOverlayComponent:import0.RendererType2 = import0.ɵcrt({
  encapsulation: 2,
  styles: styles_LightboxOverlayComponent,
  data: {}
}
);
export function View_LightboxOverlayComponent_0(l:any):import0.ɵViewDefinition {
  return import0.ɵvid(0,([] as any[]),(null as any),(null as any));
}
function View_LightboxOverlayComponent_Host_0(l:any):import0.ɵViewDefinition {
  return import0.ɵvid(0,[
      (l()(),import0.ɵeld(0,(null as any),(null as any),1,'div',[[
        'lb-overlay',
        ''
      ]
      ],[[
        8,
        'className',
        0
      ]
    ],[
      [
        (null as any),
        'click'
      ]
      ,
      [
        'window',
        'resize'
      ]

    ]
    ,(v,en,$event) => {
      var ad:boolean = true;
      if (('click' === en)) {
        const pd_0:any = ((<any>import0.ɵnov(v,1).close()) !== false);
        ad = (pd_0 && ad);
      }
      if (('window:resize' === en)) {
        const pd_1:any = ((<any>import0.ɵnov(v,1).onResize()) !== false);
        ad = (pd_1 && ad);
      }
      return ad;
    },View_LightboxOverlayComponent_0,RenderType_LightboxOverlayComponent)),
    import0.ɵdid(2187264,(null as any),0,import1.LightboxOverlayComponent,[
      import0.ElementRef,
      import0.Renderer,
      import2.LightboxEvent,
      import3.DOCUMENT
    ]
    ,(null as any),(null as any))
  ]
  ,(null as any),(ck,v) => {
    const currVal_0:any = import0.ɵnov(v,1)._classList;
    ck(v,0,0,currVal_0);
  });
}
export const LightboxOverlayComponentNgFactory:import0.ComponentFactory<import1.LightboxOverlayComponent> = import0.ɵccf('[lb-overlay]',import1.LightboxOverlayComponent,View_LightboxOverlayComponent_Host_0,{
  options: 'options',
  cmpRef: 'cmpRef'
}
,{},([] as any[]));
