// angular-2-dropdown-multiselect/esm2015/dropdown/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function IMultiSelectOption() { }
function IMultiSelectOption_tsickle_Closure_declarations() {
    /** @type {?} */
    IMultiSelectOption.prototype.id;
    /** @type {?} */
    IMultiSelectOption.prototype.name;
    /** @type {?|undefined} */
    IMultiSelectOption.prototype.disabled;
    /** @type {?|undefined} */
    IMultiSelectOption.prototype.isLabel;
    /** @type {?|undefined} */
    IMultiSelectOption.prototype.parentId;
    /** @type {?|undefined} */
    IMultiSelectOption.prototype.params;
    /** @type {?|undefined} */
    IMultiSelectOption.prototype.classes;
    /** @type {?|undefined} */
    IMultiSelectOption.prototype.image;
}
/**
 * @record
 */
export function IMultiSelectSettings() { }
function IMultiSelectSettings_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.pullRight;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.enableSearch;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.closeOnClickOutside;
    /**
     * 0 - By default
     * If `enableSearch=true` and total amount of items more then `searchRenderLimit` (0 - No limit)
     * then render items only when user typed more then or equal `searchRenderAfter` charachters
     * @type {?|undefined}
     */
    IMultiSelectSettings.prototype.searchRenderLimit;
    /**
     * 3 - By default
     * @type {?|undefined}
     */
    IMultiSelectSettings.prototype.searchRenderAfter;
    /**
     * 0 - By default
     * If >0 will render only N first items
     * @type {?|undefined}
     */
    IMultiSelectSettings.prototype.searchMaxLimit;
    /**
     * 0 - By default
     * Used with searchMaxLimit to further limit rendering for optimization
     * Should be less than searchMaxLimit to take effect
     * @type {?|undefined}
     */
    IMultiSelectSettings.prototype.searchMaxRenderedItems;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.checkedStyle;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.buttonClasses;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.itemClasses;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.containerClasses;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.selectionLimit;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.minSelectionLimit;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.closeOnSelect;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.autoUnselect;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.showCheckAll;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.showUncheckAll;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.fixedTitle;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.dynamicTitleMaxItems;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.maxHeight;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.displayAllSelectedText;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.isLazyLoad;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.loadViewDistance;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.stopScrollPropagation;
    /** @type {?|undefined} */
    IMultiSelectSettings.prototype.selectAddedValues;
    /**
     * false - By default
     * If activated label IDs don't count and won't be written to the model.
     * @type {?|undefined}
     */
    IMultiSelectSettings.prototype.ignoreLabels;
    /**
     * false - By default
     * If activated, the title will show selections in the order they were selected.
     * @type {?|undefined}
     */
    IMultiSelectSettings.prototype.maintainSelectionOrderInTitle;
    /**
     * \@default true
     * Set the focus back to the input control when the dropdown closed
     * @type {?|undefined}
     */
    IMultiSelectSettings.prototype.focusBack;
}
/**
 * @record
 */
export function IMultiSelectTexts() { }
function IMultiSelectTexts_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.checkAll;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.uncheckAll;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.checked;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.checkedPlural;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.searchPlaceholder;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.searchEmptyResult;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.searchNoRenderText;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.defaultTitle;
    /** @type {?|undefined} */
    IMultiSelectTexts.prototype.allSelected;
}
