// ng-lz-string/ng-lz-string.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('lz-string')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', 'lz-string'], factory) :
	(factory((global['ng-lz-string'] = {}),global.core,global.common,global.JSLZString));
}(this, (function (exports,core,common,JSLZString) { 'use strict';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var LZStringService = (function () {
    function LZStringService() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    LZStringService.prototype.compress = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        // TODO: Code to compress
        return JSLZString.compress(data);
    };
    /**
     * @param {?} compressed
     * @return {?}
     */
    LZStringService.prototype.decompress = /**
     * @param {?} compressed
     * @return {?}
     */
    function (compressed) {
        // TODO: Code to decompress
        return JSLZString.decompress(compressed);
    };
    LZStringService.decorators = [
        { type: core.Injectable },
    ];
    /** @nocollapse */
    LZStringService.ctorParameters = function () { return []; };
    return LZStringService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var LZStringModule = (function () {
    function LZStringModule() {
    }
    /**
     * @return {?}
     */
    LZStringModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: LZStringModule,
            providers: [LZStringService]
        };
    };
    LZStringModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule
                    ],
                    declarations: [],
                    exports: []
                },] },
    ];
    /** @nocollapse */
    LZStringModule.ctorParameters = function () { return []; };
    return LZStringModule;
}());

exports.LZStringModule = LZStringModule;
exports.LZStringService = LZStringService;

Object.defineProperty(exports, '__esModule', { value: true });

})));
